File: client/android/app/src/main/java/com/clawdbot/android/mchat/MChatChatController.kt
package com.clawdbot.android.mchat

import android.util.Log
import com.clawdbot.android.chat.ChatMessage
import com.clawdbot.android.chat.ChatMessageContent
import com.clawdbot.android.chat.ChatSessionEntry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.util.UUID

/**
 * MChat 聊天控制：会话=员工(employee_id)，发 msg.send_private，收件箱投递解析后按 from_employee_id 归入会话；员工列表来自 org.tree。
 * 支持历史消息缓存：切换回对话时显示本地缓存，收发消息时写入缓存。
 */
class MChatChatController(
  private val scope: CoroutineScope,
  private val connection: MChatConnection,
  private val historyCache: ChatHistoryCache? = null,
  initialMessagesByPeer: Map<String, List<ChatMessage>>? = null,
) {
  private val json = Json { ignoreUnknownKeys = true }

  private val _employees = MutableStateFlow<List<EmployeeEntry>>(emptyList())
  val employees: StateFlow<List<EmployeeEntry>> = _employees.asStateFlow()

  private val _messagesByPeer = MutableStateFlow<Map<String, List<ChatMessage>>>(initialMessagesByPeer ?: emptyMap())
  val messagesByPeer: StateFlow<Map<String, List<ChatMessage>>> = _messagesByPeer.asStateFlow()

  private val _currentPeerId = MutableStateFlow<String?>(null)
  val currentPeerId: StateFlow<String?> = _currentPeerId.asStateFlow()

  private val _errorText = MutableStateFlow<String?>(null)
  val errorText: StateFlow<String?> = _errorText.asStateFlow()

  private val _sessions = MutableStateFlow<List<ChatSessionEntry>>(emptyList())
  val sessions: StateFlow<List<ChatSessionEntry>> = _sessions.asStateFlow()

  init {
    (initialMessagesByPeer ?: emptyMap()).takeIf { it.isNotEmpty() }?.let { updateSessionsFromPeers(it.keys) }
  }

  data class EmployeeEntry(
    val employee_id: String,
    val name: String,
    val department_id: String?,
    val is_ai_agent: Boolean,
  )

  fun setInboxCallback() {
    // Inbox callback is set on connection; we handle in onInboxMessage which is passed from runtime
  }

  fun handleInboxMessage(payload: String) {
    try {
      val obj = json.parseToJsonElement(payload).jsonObject
      val from = obj["from_employee_id"]?.jsonPrimitive?.content ?: return
      val content = obj["content"]
      val text = when (content) {
        is kotlinx.serialization.json.JsonPrimitive -> content.content
        is kotlinx.serialization.json.JsonObject -> content["text"]?.jsonPrimitive?.content ?: content.toString()
        else -> content?.toString() ?: ""
      }
      val msgId = obj["msg_id"]?.jsonPrimitive?.content ?: UUID.randomUUID().toString()
      val sentAt = obj["sent_at"]?.jsonPrimitive?.content ?: ""
      val msg = ChatMessage(
        id = msgId,
        role = "assistant",
        content = listOf(ChatMessageContent(type = "text", text = text)),
        timestampMs = parseIsoOrNull(sentAt) ?: System.currentTimeMillis(),
      )
      scope.launch(Dispatchers.Main) {
        val map = _messagesByPeer.value.toMutableMap()
        val list = (map[from] ?: emptyList()).toMutableList()
        list.add(msg)
        map[from] = list
        _messagesByPeer.value = map
        updateSessionsFromPeers(map.keys)
        historyCache?.let { scope.launch { it.save(connection.myEmployeeId, from, list) } }
      }
    } catch (e: Throwable) {
      Log.w("MChatChat", "handleInboxMessage parse error", e)
    }
  }

  fun loadEmployees() {
    scope.launch(Dispatchers.IO) {
      _errorText.value = null
      val res = connection.request("org.tree", emptyMap())
      if (res.code != 0) {
        _errorText.value = res.message
        return@launch
      }
      val data = res.data ?: return@launch
      try {
        val root = json.parseToJsonElement(data).jsonObject
        val empArray = root["employees"]?.jsonArray ?: return@launch
        val list = empArray.mapNotNull { el ->
          val o = el.jsonObject
          EmployeeEntry(
            employee_id = o["employee_id"]?.jsonPrimitive?.content ?: return@mapNotNull null,
            name = o["name"]?.jsonPrimitive?.content ?: "",
            department_id = o["department_id"]?.jsonPrimitive?.content,
            is_ai_agent = o["is_ai_agent"]?.jsonPrimitive?.content == "true",
          )
        }
        _employees.value = list
        updateSessionsFromPeers(_messagesByPeer.value.keys)
      } catch (e: Throwable) {
        _errorText.value = e.message
      }
    }
  }

  fun selectPeer(employeeId: String?) {
    _currentPeerId.value = employeeId
  }

  fun sendMessage(peerEmployeeId: String, text: String) {
    val trimmed = text.trim()
    if (trimmed.isEmpty()) return
    scope.launch(Dispatchers.IO) {
      _errorText.value = null
      val optimisticId = UUID.randomUUID().toString()
      val optimistic = ChatMessage(
        id = optimisticId,
        role = "user",
        content = listOf(ChatMessageContent(type = "text", text = trimmed)),
        timestampMs = System.currentTimeMillis(),
      )
      scope.launch(Dispatchers.Main) {
        val map = _messagesByPeer.value.toMutableMap()
        val list = (map[peerEmployeeId] ?: emptyList()).toMutableList()
        list.add(optimistic)
        map[peerEmployeeId] = list
        _messagesByPeer.value = map
        updateSessionsFromPeers(map.keys)
        historyCache?.let { scope.launch { it.save(connection.myEmployeeId, peerEmployeeId, list) } }
      }
      val res = connection.request("msg.send_private", mapOf(
        "to_employee_id" to peerEmployeeId,
        "content" to trimmed,
      ))
      if (res.code != 0) {
        _errorText.value = res.message
      }
    }
  }

  fun getEmployee(employeeId: String, onResult: (EmployeeEntry?) -> Unit) {
    scope.launch(Dispatchers.IO) {
      val res = connection.request("employee.get", mapOf("employee_id" to employeeId))
      if (res.code != 0) {
        scope.launch(Dispatchers.Main) { onResult(null) }
        return@launch
      }
      val data = res.data ?: run {
        scope.launch(Dispatchers.Main) { onResult(null) }
        return@launch
      }
      try {
        val o = json.parseToJsonElement(data).jsonObject
        val entry = EmployeeEntry(
          employee_id = o["employee_id"]?.jsonPrimitive?.content ?: employeeId,
          name = o["name"]?.jsonPrimitive?.content ?: "",
          department_id = o["department_id"]?.jsonPrimitive?.content,
          is_ai_agent = o["is_ai_agent"]?.jsonPrimitive?.content == "true",
        )
        scope.launch(Dispatchers.Main) { onResult(entry) }
      } catch (_: Throwable) {
        scope.launch(Dispatchers.Main) { onResult(null) }
      }
    }
  }

  fun currentMessages(): List<ChatMessage> {
    val peer = _currentPeerId.value ?: return emptyList()
    return _messagesByPeer.value[peer] ?: emptyList()
  }

  private fun updateSessionsFromPeers(peerIds: Set<String>) {
    val employeesMap = _employees.value.associateBy { it.employee_id }
    _sessions.value = peerIds.map { id ->
      val name = employeesMap[id]?.name ?: id
      ChatSessionEntry(key = id, updatedAtMs = null, displayName = name)
    }.sortedByDescending { (_, u, _) -> u ?: 0L }
  }

  private fun parseIsoOrNull(s: String): Long? {
    return try {
      java.time.Instant.parse(s).toEpochMilli()
    } catch (_: Throwable) {
      null
    }
  }
}

File: client/ios/MoltChat/MChat/MChatChatController.swift
import Foundation

/// MChat 聊天控制器，对标 Android MChatChatController
@MainActor
final class MChatChatController {
    private let connection: MChatConnection
    private let historyCache: ChatHistoryCache?
    
    private(set) var employees: [EmployeeEntry] = []
    private(set) var messagesByPeer: [String: [ChatMessage]] = [:]
    private(set) var currentPeerId: String?
    private(set) var errorText: String?
    private(set) var sessions: [ChatSessionEntry] = []
    
    var onEmployeesChanged: (([EmployeeEntry]) -> Void)?
    var onMessagesChanged: (([String: [ChatMessage]]) -> Void)?
    var onCurrentPeerChanged: ((String?) -> Void)?
    var onErrorChanged: ((String?) -> Void)?
    var onSessionsChanged: (([ChatSessionEntry]) -> Void)?
    
    init(connection: MChatConnection, historyCache: ChatHistoryCache?, initialMessagesByPeer: [String: [ChatMessage]]?) {
        self.connection = connection
        self.historyCache = historyCache
        if let initial = initialMessagesByPeer {
            self.messagesByPeer = initial
            updateSessionsFromPeers(Set(initial.keys))
        }
    }
    
    func setInboxCallback(_ callback: @escaping (String) -> Void) {
        connection.setInboxCallback(callback)
    }
    
    func handleInboxMessage(_ payload: String) {
        guard let data = payload.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let from = obj["from_employee_id"] as? String else { return }
        var text = ""
        if let content = obj["content"] {
            if let s = content as? String { text = s }
            else if let o = content as? [String: Any], let t = o["text"] as? String { text = t }
        }
        let msgId = obj["msg_id"] as? String ?? UUID().uuidString
        let sentAt = obj["sent_at"] as? String ?? ""
        let timestampMs = parseIso(sentAt) ?? Int64(Date().timeIntervalSince1970 * 1000)
        let msg = ChatMessage(
            id: msgId,
            role: "assistant",
            content: [ChatMessageContent(type: "text", text: text)],
            timestampMs: timestampMs
        )
        var map = messagesByPeer
        var list = map[from] ?? []
        list.append(msg)
        map[from] = list
        messagesByPeer = map
        onMessagesChanged?(map)
        updateSessionsFromPeers(Set(map.keys))
        if let cache = historyCache {
            Task { await cache.save(myEmployeeId: connection.myEmployeeId, peerId: from, messages: list) }
        }
    }
    
    func loadEmployees() {
        errorText = nil
        onErrorChanged?(nil)
        Task {
            let res = await connection.request(action: "org.tree", params: [:])
            if res.code != 0 {
                errorText = res.message
                onErrorChanged?(res.message)
                return
            }
            guard let dataStr = res.data, let data = dataStr.data(using: .utf8),
                  let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let empArray = root["employees"] as? [[String: Any]] else { return }
            var list: [EmployeeEntry] = []
            for o in empArray {
                guard let eid = o["employee_id"] as? String else { continue }
                list.append(EmployeeEntry(
                    employeeId: eid,
                    name: o["name"] as? String ?? "",
                    departmentId: o["department_id"] as? String,
                    isAiAgent: (o["is_ai_agent"] as? String) == "true"
                ))
            }
            employees = list
            onEmployeesChanged?(list)
            updateSessionsFromPeers(Set(messagesByPeer.keys))
        }
    }
    
    func selectPeer(_ employeeId: String?) {
        currentPeerId = employeeId
        onCurrentPeerChanged?(employeeId)
    }
    
    func sendMessage(peerEmployeeId: String, text: String) {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        errorText = nil
        onErrorChanged?(nil)
        let optimistic = ChatMessage(
            id: UUID().uuidString,
            role: "user",
            content: [ChatMessageContent(type: "text", text: trimmed)],
            timestampMs: Int64(Date().timeIntervalSince1970 * 1000)
        )
        var map = messagesByPeer
        var list = map[peerEmployeeId] ?? []
        list.append(optimistic)
        map[peerEmployeeId] = list
        messagesByPeer = map
        onMessagesChanged?(map)
        updateSessionsFromPeers(Set(map.keys))
        if let cache = historyCache {
            Task { await cache.save(myEmployeeId: connection.myEmployeeId, peerId: peerEmployeeId, messages: list) }
        }
        Task {
            let res = await connection.request(action: "msg.send_private", params: [
                "to_employee_id": peerEmployeeId,
                "content": trimmed
            ])
            if res.code != 0 {
                errorText = res.message
                onErrorChanged?(res.message)
            }
        }
    }
    
    func getEmployee(employeeId: String) async -> EmployeeEntry? {
        let res = await connection.request(action: "employee.get", params: ["employee_id": employeeId])
        guard res.code == 0, let dataStr = res.data, let data = dataStr.data(using: .utf8),
              let o = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        return EmployeeEntry(
            employeeId: o["employee_id"] as? String ?? employeeId,
            name: o["name"] as? String ?? "",
            departmentId: o["department_id"] as? String,
            isAiAgent: (o["is_ai_agent"] as? String) == "true"
        )
    }
    
    func currentMessages() -> [ChatMessage] {
        guard let peer = currentPeerId else { return [] }
        return messagesByPeer[peer] ?? []
    }
    
    private func updateSessionsFromPeers(_ peerIds: Set<String>) {
        let empMap = Dictionary(uniqueKeysWithValues: employees.map { ($0.employeeId, $0.name) })
        sessions = peerIds.map { id in
            ChatSessionEntry(key: id, updatedAtMs: nil, displayName: empMap[id] ?? id)
        }
    }
    
    private func parseIso(_ s: String) -> Int64? {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = formatter.date(from: s) {
            return Int64(d.timeIntervalSince1970 * 1000)
        }
        formatter.formatOptions = [.withInternetDateTime]
        if let d = formatter.date(from: s) {
            return Int64(d.timeIntervalSince1970 * 1000)
        }
        return nil
    }
}
